### ISSUE TITLE
Ship sftp-server with sshd

### ISSUE THREAD

**Issue by alethiophile** [koreader/koreader#4104]
* KOReader version: v2015.11-1788-g00cfc46_2018-07-23
* Device: Kobo Aura One

#### Issue
SSH server doesn't provide sftp

#### Steps to reproduce
1. Start SSH server
2. Attempt `sftp root@ip`, watch it fail

It would be good to provide a /usr/libexec/sftp-server along with the included dropbear sshd. As it stands the only way to transfer files over the SSH link is output redirection and cat (`scp` doesn't work either).

There are various projects providing builds of OpenSSH's sftp-server for Kobo; one of these should be fine, or could build specially.

**Comment by NiLuJe** [comment]
We'd need to build OpenSSH ourselves, we don't bundle binaries.

**Comment by NiLuJe** [comment]
I'm interested to know how scp fails though, because it shouldn't.

**Comment by NiLuJe** [comment]
Also, I'm not terribly keen on passing a relative path to execv (and I'n not even sure that actually works), and we can't pass an absolute one, because it's defined at compile-time, and we support being installed in a number of different paths.

**Comment by NiLuJe** [comment]
TL;DR: To make it clearer: I'm inclined to #wontfix the sftp thing for a number of reasons.

The scp thing *is* interesting though, and warrants a fix.

**Comment by alethiophile** [comment]
The error I get for scp is 'sh: scp: not found', which originates from the busybox sh on the device.

Also, the path for sftp-server is always `/usr/libexec/sftp-server`, at least that I can find. I'm not sure this is even configurable at all, at least in dropbear. It doesn't seem it should be related to where the full koreader installation is.

**Comment by NiLuJe** [comment]
It is configurable, at compile-time ;).

And  we won't ever drop anything outside of onboard, because hijacking the rootfs is out of question, for a shitload of good reasons.

So we're left with putting us somewhere we can keep track of it, meaning under our own directory tree, meaning a relative path.

**Comment by NiLuJe** [comment]
And indeed it requires an scp client on the host's side, and it needs to be in PATH, so, yeaaaaah, big nope.

**Comment by NiLuJe** [comment]
As for tradeoffs, besides, err, basically scrapping the plugin and starting from scratch ;D.

I don't have patches for passwordless entry with OpenSSH, so, that's out, at the very least.

Other than that, the biggest hurdle is OpenSSL, and we already build that, so, yay.

Pubkey authentication needs a good number of whack on the head to handle vfat and/or relaxed permissions checks, but, that, I have on hand.

**Comment by NiLuJe** [comment]
... and, double-checking my own build scripts, it also requires absolute paths at build-time to know where its configs are stored, some of which are mandatory for a successful connection :D.

**Comment by NiLuJe** [comment]
I'll try to mangle dropbear into submission, so we can at least geti one of scp or sftp working, because having none is kind of a bummer.

**Comment by NiLuJe** [comment]
Landed in 1793.

As mentioned, will only behave if installed in the default location. (And essentially breaks it for people who *did* have scp/sfpt in their path, but that means they alkready have an SSH access ;)).

**Comment by thomasrebele** [comment]
I found a workaround for my device which does not support scp / sftp: the workaround uses ssh, tar, and lsyncd. I added a reference in the wiki: https://github.com/koreader/koreader/wiki/SSH

**Comment by Frenzie** [comment]
Thanks for sharing!
